onnxruntime/core/providers/cuda/tensor/scatter_nd.cc

```diff
@@ -85,5 +101,88 @@ Status ScatterND::ComputeInternal(OpKernelContext* context) const {
   return Status::OK();
 }
 
+Status ScatterNDWithAtomicReduction::ComputeInternal(OpKernelContext* context) const {
```

Comment:
I am seeing these two statements in the spec for the latest version. If we are able to validate indecies for uniqueness, then we can avoid atomics. Atomics are not free.

`The order of iteration in the above loop is not specified. In particular, indices should not have duplicate entries: that is, if idx1 != idx2, then indices[idx1] != indices[idx2]. This ensures that the output value does not depend on the iteration order.`

`In cases where reduction is set to "none", indices should not have duplicate entries: that is, if idx1 != idx2, then indices[idx1] != indices[idx2].`

Reply:
I know the implementation I did is not the best one. I took it from another Scatter operator in onnxruntime because it seemed ok for the first implementation. I'm currently working on a new one but it won't be ready before the release. I suggest keeping this one for now.

Reply:
What does Torch do? We may loose some perf grounds if they do not do atomics. Also, we can detect, if indicies are unique, we can forgo atomics.

Reply:
I'm looking into an implementation without atomics. The version I have can be 4 times faster or 2 times slower. It depends on the input size. I did not check if that was the only implementation but torch is also using atomic in put_kernel (through a method called fastSpecializedAtomicAdd).